net.sourceforge.pmd.eclipse.plugin/src/main/java/net/sourceforge/pmd/eclipse/runtime/preferences/impl/PreferencesManagerImpl.java

```diff
@@ -637,15 +638,17 @@ private void updateConfiguredProjects(RuleSet updatedRuleSet) {
             if (project.isAccessible()) {
                 try {
                     IProjectProperties properties = PMDPlugin.getDefault().loadProjectProperties(project);
-                    RuleSet projectRuleSet = properties.getProjectRuleSet();
+                    RuleSets projectRuleSet = properties.getProjectRuleSets();
                     if (projectRuleSet != null) {
-                        projectRuleSet = RuleSetUtil.addRules(projectRuleSet, getNewRules(updatedRuleSet));
-                        projectRuleSet = RuleSetUtil.setExcludePatterns(projectRuleSet,
-                                updatedRuleSet.getExcludePatterns());
-                        projectRuleSet = RuleSetUtil.setIncludePatterns(projectRuleSet,
-                                updatedRuleSet.getIncludePatterns());
-                        properties.setProjectRuleSet(projectRuleSet);
-                        properties.sync();
+                      projectRuleSet.getAllRules().addAll(getNewRules(updatedRuleSet));
+                      for(RuleSet rs : projectRuleSet.getAllRuleSets()) {
+                        rs = RuleSetUtil.setExcludePatterns(rs,
```

Comment:
The RuleSetUtil will return a copy of the original ruleset with the changed properties... so, in the end, we need to collect the rulesets again... e.g.

```suggestion
                    RuleSets newRuleSets = new RuleSets();
                    for (RuleSet rs : projectRuleSets.getAllRuleSets()) {
                        rs = RuleSetUtil.setExcludePatterns(rs, updatedRuleSet.getExcludePatterns());
                        rs = RuleSetUtil.setIncludePatterns(rs, updatedRuleSet.getIncludePatterns());
                        newRuleSets.addRuleSet(rs);
                    }
                    properties.setProjectRuleSets(newRuleSets);
                    properties.sync();
```

Reply:
I see. I was thinking it modified the original.